File: scripts/deploy_helios.py
#!/usr/bin/env python3
"""
Helios full deployment pipeline — pure Python, no casper-client needed.

Steps:
  1. Generate keys (if not already done)
  2. Print faucet instructions and wait for confirmation
  3. Deploy OracleRegistry
  4. Deploy DataMarket (with registry hash)
  5. Wire them: call set_market on OracleRegistry
  6. Deploy FundVault
  7. Deploy Governance
  8. Wire FundVault → Governance
  9. Write agents/testnet.env
  10. Produce first on-chain activity (register + attest + anchor)
  11. Print explorer links for all transactions

Usage:
  python3 scripts/deploy_helios.py
  python3 scripts/deploy_helios.py --keys-dir keys --node https://rpc.testnet.cspr.cloud
  python3 scripts/deploy_helios.py --skip-keygen   # if keys already exist
"""

from __future__ import annotations
import argparse, json, os, sys, time
from pathlib import Path

# Add project root so we can import casper_deploy
sys.path.insert(0, str(Path(__file__).parent))
from casper_deploy import (
    CasperKey,
    generate_key,
    install_wasm,
    call_entry_point,
    wait_for_deploy,
    arg_string,
    arg_u64,
    arg_u32,
    arg_u512,
    _rpc,
    NODES,
    EXPLORER,
)

ROOT = Path(__file__).parent.parent
WASM = ROOT / "contracts" / "wasm"
AGENTS = ROOT / "agents"

# ── helpers ───────────────────────────────────────────────────────────────────


def wait_and_get_contract_hash(deploy_hash: str, contract_name: str, node: str) -> str:
    """Wait for deploy, then extract contract hash from execution result."""
    wait_for_deploy(deploy_hash, node=node)
    time.sleep(2)  # allow state to propagate
    try:
        result = _rpc(
            "info_get_transaction", {"transaction_hash": {"Deploy": deploy_hash}}, node
        )
        txn = result.get("transaction", result.get("deploy", {}))
        # execution_info path (Casper 2.x)
        exec_info = txn.get("execution_info", {})
        if exec_info:
            effects = exec_info.get("execution_result", {}).get("effects", {})
            for transform in effects.get("transforms", []):
                if transform.get("kind", {}).get("WriteContract") is not None:
                    return transform["key"].replace("hash-", "")
        # fallback: user reads from cspr.live
        print(f"  ⚠ Could not auto-extract {contract_name} hash.")
        print(f"    Open {EXPLORER}/deploy/{deploy_hash}")
        print(
            f"    Find the 'WriteContract' entry → copy the hash (without 'hash-' prefix)"
        )
        return input(f"  Paste {contract_name} hash: ").strip().replace("hash-", "")
    except Exception as e:
        print(f"  ⚠ Hash extraction error ({e}), please paste manually:")
        print(f"    Explorer: {EXPLORER}/deploy/{deploy_hash}")
        return input(f"  Paste {contract_name} hash: ").strip().replace("hash-", "")


# ── main deployment flow ──────────────────────────────────────────────────────


def main():
    parser = argparse.ArgumentParser(description="Deploy Helios to Casper Testnet")
    parser.add_argument("--keys-dir", default="keys")
    parser.add_argument("--node", default=NODES[0])
    parser.add_argument("--skip-keygen", action="store_true")
    parser.add_argument(
        "--skip-activity",
        action="store_true",
        help="Skip the post-deploy on-chain activity step",
    )
    parser.add_argument(
        "--resume",
        type=str,
        default=None,
        metavar="testnet.env",
        help="Resume from an existing testnet.env (skip deploys)",
    )
    args = parser.parse_args()

    keys_dir = ROOT / args.keys_dir
    node = args.node

    print("═" * 60)
    print("  Helios Testnet Deployment  (pure Python)")
    print("═" * 60)

    # ── Step 0: check node ────────────────────────────────────────────────────
    print("\n[0] Checking node connectivity…")
    try:
        status = _rpc("info_get_status", {}, node)
        print(f"    ✓ {node} — chain: {status.get('chainspec_name', '?')}")
    except Exception as e:
        print(f"    ✗ Cannot reach {node}: {e}")
        for alt in NODES[1:]:
            try:
                _rpc("info_get_status", {}, alt)
                print(f"    ✓ Falling back to {alt}")
                node = alt
                break
            except Exception:
                pass
        else:
            print("    ✗ All nodes unreachable. Check your connection.")
            sys.exit(1)

    # ── Step 1: keygen ────────────────────────────────────────────────────────
    roles = [
        "oracle_tbill",
        "oracle_gold",
        "oracle_reindex",
        "fund_agent",
        "risk_agent",
    ]
    if args.resume:
        print(f"\n[1] Resuming — loading keys from {keys_dir}")
    elif not args.skip_keygen:
        print(f"\n[1] Generating keypairs in {keys_dir}/")
        keys_dir.mkdir(parents=True, exist_ok=True)
        for role in roles:
            rdir = keys_dir / role
            key_path = rdir / "secret_key.pem"
            if key_path.exists():
                print(f"    {role}: already exists — skipping")
                continue
            rdir.mkdir(parents=True, exist_ok=True)
            k = generate_key(str(key_path))
            print(f"    {role}: {k.account_hash()}")
    else:
        print(f"\n[1] Skipping keygen (--skip-keygen)")

    keys = {r: CasperKey(str(keys_dir / r / "secret_key.pem")) for r in roles}
    deployer_key = keys["fund_agent"]

    # ── Faucet prompt ─────────────────────────────────────────────────────────
    if not args.resume:
        print("\n[2] Fund accounts via faucet BEFORE deploying")
        print("    URL: https://testnet.cspr.live/tools/faucet")
        print("    Each account needs ≥ 1000 test CSPR. Accounts:\n")
        for role, key in keys.items():
            print(f"    {role:20s}  {key.pubkey_hex()}")
        print()
        input("    Press ENTER when all accounts are funded… ")

    # ── Wasm check ────────────────────────────────────────────────────────────
    if not args.resume:
        needed = [
            "OracleRegistry.wasm",
            "DataMarket.wasm",
            "FundVault.wasm",
            "Governance.wasm",
        ]
        missing_wasm = [w for w in needed if not (WASM / w).exists()]
        if missing_wasm:
            print(f"\n✗ Missing wasm files: {missing_wasm}")
            print(f"  Run:  bash scripts/build_contracts.sh")
            sys.exit(1)
        print("\n[3] WASM files found ✓")
        # Verify exports
        sys.path.insert(0, str(ROOT / "scripts"))
        import check_wasm_exports

        bad = 0
        for w in needed:
            errors, func_names, has_mem = check_wasm_exports.check_file(str(WASM / w))
            if errors:
                print(f"  ✗ {w}: {errors[0]}")
                bad += 1
            else:
                print(f"  ✓ {w}")
        if bad:
            print("  Fix build first: bash scripts/build_contracts.sh")
            sys.exit(1)

    # ── Deploy or resume ──────────────────────────────────────────────────────
    if args.resume:
        env = {}
        for line in Path(args.resume).read_text().splitlines():
            if "=" in line and not line.startswith("#"):
                k, v = line.split("=", 1)
                env[k.strip()] = v.strip()
        registry_hash = env.get("REGISTRY_HASH", "")
        market_hash = env.get("MARKET_HASH", "")
        vault_hash = env.get("VAULT_HASH", "")
        gov_hash = env.get("GOV_HASH", "")
        print(f"\n[3-7] Resuming from existing env:")
        print(f"    REGISTRY: {registry_hash}")
        print(f"    MARKET:   {market_hash}")
        print(f"    VAULT:    {vault_hash}")
        print(f"    GOV:      {gov_hash}")
    else:
        txns = {}

        print("\n[4] Deploying OracleRegistry…")
        h = install_wasm(deployer_key, str(WASM / "OracleRegistry.wasm"), node=node)
        print(f"    deploy: {h}\n    explorer: {EXPLORER}/deploy/{h}")
        txns["registry_deploy"] = h
        registry_hash = wait_and_get_contract_hash(h, "OracleRegistry", node)
        print(f"    OracleRegistry hash: {registry_hash}")

        print("\n[5] Deploying DataMarket…")
        h = install_wasm(
            deployer_key,
            str(WASM / "DataMarket.wasm"),
            named_args=[
                arg_string("registry_hash", registry_hash),
                arg_u32("fee_bps", 250),
            ],
            node=node,
        )
        print(f"    deploy: {h}\n    explorer: {EXPLORER}/deploy/{h}")
        txns["market_deploy"] = h
        market_hash = wait_and_get_contract_hash(h, "DataMarket", node)
        print(f"    DataMarket hash: {market_hash}")

        print("\n[5b] Wiring: OracleRegistry.set_market…")
        h = call_entry_point(
            deployer_key,
            registry_hash,
            "set_market",
            [arg_string("market", market_hash)],
            node=node,
        )
        wait_for_deploy(h, node=node)
        txns["set_market"] = h
        print(f"    wired ✓  deploy: {h}")

        print("\n[6] Deploying FundVault…")
        fund_acct = deployer_key.account_hash()
        h = install_wasm(
            deployer_key,
            str(WASM / "FundVault.wasm"),
            named_args=[
                arg_string("operator", fund_acct),
                arg_string("governance_hash", "pending"),  # updated after step 7
            ],
            node=node,
        )
        print(f"    deploy: {h}\n    explorer: {EXPLORER}/deploy/{h}")
        txns["vault_deploy"] = h
        vault_hash = wait_and_get_contract_hash(h, "FundVault", node)
        print(f"    FundVault hash: {vault_hash}")

        print("\n[7] Deploying Governance…")
        risk_acct = keys["risk_agent"].account_hash()
        h = install_wasm(
            deployer_key,
            str(WASM / "Governance.wasm"),
            named_args=[
                arg_string("proposer", fund_acct),
                arg_string("risk_agent", risk_acct),
                arg_u64("veto_window_ms", 90_000),
            ],
            node=node,
        )
        print(f"    deploy: {h}\n    explorer: {EXPLORER}/deploy/{h}")
        txns["gov_deploy"] = h
        gov_hash = wait_and_get_contract_hash(h, "Governance", node)
        print(f"    Governance hash: {gov_hash}")

        print("\n[7b] Wiring: FundVault.set_governance…")
        h = call_entry_point(
            deployer_key,
            vault_hash,
            "set_governance",
            [arg_string("governance_hash", gov_hash)],
            node=node,
        )
        wait_for_deploy(h, node=node)
        txns["set_governance"] = h
        print(f"    wired ✓  deploy: {h}")

        # save all deploy hashes for reference
        Path("deploy_hashes.json").write_text(json.dumps(txns, indent=2))
        print(f"\n    All deploy hashes saved to deploy_hashes.json")

    # ── Write testnet.env ──────────────────────────────────────────────────────
    env_path = AGENTS / "testnet.env"
    env_content = f"""# Helios Testnet Config — generated by deploy_helios.py
ORACLE_TBILL_KEY={keys_dir}/oracle_tbill/secret_key.pem
ORACLE_GOLD_KEY={keys_dir}/oracle_gold/secret_key.pem
ORACLE_REINDEX_KEY={keys_dir}/oracle_reindex/secret_key.pem
FUND_AGENT_KEY={keys_dir}/fund_agent/secret_key.pem
RISK_AGENT_KEY={keys_dir}/risk_agent/secret_key.pem
REGISTRY_HASH={registry_hash}
MARKET_HASH={market_hash}
VAULT_HASH={vault_hash}
GOV_HASH={gov_hash}
"""
    env_path.write_text(env_content)
    print(f"\n[8] agents/testnet.env written ✓")

    # ── On-chain activity ──────────────────────────────────────────────────────
    if not args.skip_activity:
        print("\n[9] Producing on-chain activity (register + attest + anchor)…")
        activity_txns = []

        for role, feed_key, title, price in [
            ("oracle_tbill", "us_tbill_3m", "US T-Bill 3M Yield", 2_000_000_000),
            ("oracle_gold", "gold_spot_usd", "Gold Spot Price (USD)", 3_000_000_000),
            ("oracle_reindex", "cn_reindex_chz", "Zhongshan RE Index", 5_000_000_000),
        ]:
            k = keys[role]
            acct = k.account_hash()
            print(f"\n  Registering {role}…")
            h = call_entry_point(
                k,
                registry_hash,
                "register",
                [
                    arg_string("name", title),
                    arg_string("category", "rwa"),
                    arg_string("endpoint", f"https://helios.example/{feed_key}"),
                    arg_u64("price_motes", price),
                ],
                node=node,
            )
            wait_for_deploy(h, node=node)
            activity_txns.append({"action": f"register:{role}", "hash": h})
            print(f"    register: {EXPLORER}/deploy/{h}")

            print(f"  Posting attestation for {feed_key}…")
            h = call_entry_point(
                k,
                registry_hash,
                "post_attestation",
                [
                    arg_string("feed_key", feed_key),
                    arg_string("value", "42.0"),
                ],
                node=node,
            )
            wait_for_deploy(h, node=node)
            activity_txns.append({"action": f"attest:{feed_key}", "hash": h})
            print(f"    attest: {EXPLORER}/deploy/{h}")

        Path("activity_hashes.json").write_text(json.dumps(activity_txns, indent=2))
        print(f"\n  Activity hashes saved to activity_hashes.json")

    # ── Summary ────────────────────────────────────────────────────────────────
    print("\n" + "═" * 60)
    print("  DEPLOYMENT COMPLETE")
    print("═" * 60)
    print(f"  OracleRegistry : {EXPLORER}/contract/{registry_hash}")
    print(f"  DataMarket     : {EXPLORER}/contract/{market_hash}")
    print(f"  FundVault      : {EXPLORER}/contract/{vault_hash}")
    print(f"  Governance     : {EXPLORER}/contract/{gov_hash}")
    print()
    print("  Next: run the testnet agent loop")
    print("    export HELIOS_MODE=testnet")
    print("    python3 scripts/testnet_round.py --rounds 3")
    print()


if __name__ == "__main__":
    main()

File: scripts/deploy_helios.js
#!/usr/bin/env node
/**
 * Helios Testnet Deployer — Casper official JS SDK
 *
 * Prerequisites:
 *   node >= 18
 *   npm install casper-js-sdk
 *
 * Usage:
 *   node scripts/deploy_helios.js keygen
 *   node scripts/deploy_helios.js status
 *   node scripts/deploy_helios.js install \
 *       --key "keys/fund_agent/secret_key.pem" \
 *       --wasm contracts/wasm/OracleRegistry.wasm
 *   node scripts/deploy_helios.js call \
 *       --key "keys/oracle_tbill/secret_key.pem" \
 *       --contract <HASH> \
 *       --entry-point register \
 *       --args "name:string=TBill Oracle" "category:string=rwa" \
 *               "endpoint:string=http://localhost:8451/quote" "price_motes:u64=2000000000"
 *   node scripts/deploy_helios.js deploy-all
 */

const fs   = require("fs");
const path = require("path");

// ── Lazy-load SDK so `node deploy_helios.js` gives a clean error if not installed
let SDK;
function sdk() {
  if (!SDK) {
    try {
      SDK = require("casper-js-sdk");
    } catch {
      console.error("casper-js-sdk not installed. Run:  npm install casper-js-sdk");
      process.exit(1);
    }
  }
  return SDK;
}

const ROOT     = path.resolve(__dirname, "..");
const WASM_DIR = path.join(ROOT, "contracts", "wasm");
const KEYS_DIR = path.join(ROOT, "keys");
const AGENTS   = path.join(ROOT, "agents");

const NODE_URL  = process.env.CASPER_NODE || "https://rpc.testnet.cspr.cloud/rpc";
const CHAIN     = "casper-test";
const EXPLORER  = "https://testnet.cspr.live";
const GAS_INSTALL = "400000000000";   // 400 CSPR for wasm install
const GAS_CALL    = "5000000000";     // 5 CSPR for entry-point call

// ── Key helpers ────────────────────────────────────────────────────────────────

function loadKey(keyPath) {
  const { Keys } = sdk();
  return Keys.Ed25519.loadKeyPairFromPrivateFile(keyPath);
}

function keygen(outDir) {
  const { Keys } = sdk();
  fs.mkdirSync(outDir, { recursive: true });
  const kp = Keys.Ed25519.new();
  const secretPath = path.join(outDir, "secret_key.pem");
  const pubPath    = path.join(outDir, "public_key.pem");
  const hexPath    = path.join(outDir, "public_key_hex");
  fs.writeFileSync(secretPath, kp.exportPrivateKeyInPem());
  fs.writeFileSync(pubPath,    kp.exportPublicKeyInPem());
  fs.writeFileSync(hexPath,    kp.publicKey.toHex());
  return kp;
}

// ── RPC helpers ────────────────────────────────────────────────────────────────

async function rpc(method, params) {
  const { CasperServiceByJsonRPC } = sdk();
  const client = new CasperServiceByJsonRPC(NODE_URL);
  return client[method] ? client[method](...Object.values(params))
                        : Promise.reject(new Error(`Unknown method: ${method}`));
}

async function getClient() {
  const { CasperServiceByJsonRPC } = sdk();
  return new CasperServiceByJsonRPC(NODE_URL);
}

async function waitForDeploy(deployHash, timeoutMs = 300_000) {
  const client = await getClient();
  process.stdout.write(`   waiting for ${deployHash.slice(0,16)}…`);
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    try {
      const result = await client.getDeployInfo(deployHash);
      if (result?.execution_results?.length > 0) {
        const er = result.execution_results[0].result;
        if (er.Failure) {
          console.log(" FAILED");
          throw new Error(`Deploy failed: ${JSON.stringify(er.Failure.error_message)}`);
        }
        console.log(" ✓");
        return result;
      }
    } catch (e) {
      if (e.message?.includes("Deploy failed")) throw e;
    }
    process.stdout.write(".");
    await new Promise(r => setTimeout(r, 4000));
  }
  console.log(" TIMEOUT");
  throw new Error(`Deploy ${deployHash} not finalised in ${timeoutMs / 1000}s`);
}

// ── Arg builders ───────────────────────────────────────────────────────────────

function parseArgs(rawArgs) {
  const { RuntimeArgs, CLValueBuilder, CLTypeTag } = sdk();
  const rArgs = RuntimeArgs.fromMap({});
  for (const a of (rawArgs || [])) {
    const colonIdx = a.indexOf(":");
    const eqIdx    = a.indexOf("=", colonIdx);
    const name = a.slice(0, colonIdx);
    const typ  = a.slice(colonIdx + 1, eqIdx);
    const val  = a.slice(eqIdx + 1);
    switch (typ) {
      case "string": rArgs.insert(name, CLValueBuilder.string(val));          break;
      case "u64":    rArgs.insert(name, CLValueBuilder.u64(BigInt(val)));      break;
      case "u32":    rArgs.insert(name, CLValueBuilder.u32(Number(val)));      break;
      case "u512":   rArgs.insert(name, CLValueBuilder.u512(BigInt(val)));     break;
      case "bool":   rArgs.insert(name, CLValueBuilder.bool(
                       val === "true" || val === "1" || val === "yes"));       break;
      default: throw new Error(`Unknown arg type '${typ}' in '${a}'`);
    }
  }
  return rArgs;
}

// ── Deploy functions ───────────────────────────────────────────────────────────

async function installWasm(keyPair, wasmPath, rawArgs, paymentMotes) {
  const { DeployUtil, RuntimeArgs } = sdk();
  const client = await getClient();
  const wasm   = new Uint8Array(fs.readFileSync(wasmPath));
  const args   = parseArgs(rawArgs);

  const deploy = DeployUtil.makeDeploy(
    new DeployUtil.DeployHeader(
      keyPair.publicKey,
      Date.now(),
      1_800_000,      // TTL 30 min
      1,              // gas price
      [],
      CHAIN,
    ),
    DeployUtil.ExecutableDeployItem.newModuleBytes(wasm, args),
    DeployUtil.standardPayment(paymentMotes || GAS_INSTALL),
  );
  const signed = DeployUtil.signDeploy(deploy, keyPair);
  const result = await client.deploy(signed);
  return result.deploy_hash;
}

async function callContract(keyPair, contractHash, entryPoint, rawArgs, paymentMotes) {
  const { DeployUtil, CLValueBuilder, contracts } = sdk();
  const client = await getClient();
  const args   = parseArgs(rawArgs);

  const deploy = DeployUtil.makeDeploy(
    new DeployUtil.DeployHeader(
      keyPair.publicKey,
      Date.now(),
      1_800_000,
      1,
      [],
      CHAIN,
    ),
    DeployUtil.ExecutableDeployItem.newStoredContractByHash(
      contractHash, entryPoint, args),
    DeployUtil.standardPayment(paymentMotes || GAS_CALL),
  );
  const signed = DeployUtil.signDeploy(deploy, keyPair);
  const result = await client.deploy(signed);
  return result.deploy_hash;
}

// ── Contract hash extractor ────────────────────────────────────────────────────

async function extractContractHash(deployHash) {
  const client = await getClient();
  await new Promise(r => setTimeout(r, 2000));
  try {
    const info = await client.getDeployInfo(deployHash);
    const transforms = info?.execution_results?.[0]
      ?.result?.Success?.effect?.transforms || [];
    for (const t of transforms) {
      if (t.transform?.WriteContract || t.transform?.WriteContractWasm) {
        return t.key.replace("hash-", "");
      }
    }
  } catch {}
  // fallback: prompt user
  console.log(`  Open: ${EXPLORER}/deploy/${deployHash}`);
  console.log("  Find 'WriteContract' in the execution effects → copy the hash");
  const readline = require("readline");
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise(resolve => {
    rl.question("  Paste contract hash (without 'hash-' prefix): ", ans => {
      rl.close();
      resolve(ans.trim().replace(/^hash-/, ""));
    });
  });
}

// ── Full deployment flow ───────────────────────────────────────────────────────

async function deployAll() {
  const roles = ["oracle_tbill","oracle_gold","oracle_reindex","fund_agent","risk_agent"];

  // Step 1: ensure keys exist
  console.log("\n[1] Loading / generating keypairs…");
  const keys = {};
  for (const role of roles) {
    const dir = path.join(KEYS_DIR, role);
    const kp  = path.join(dir, "secret_key.pem");
    if (fs.existsSync(kp)) {
      keys[role] = loadKey(kp);
      console.log(`    ${role}: loaded`);
    } else {
      fs.mkdirSync(dir, { recursive: true });
      keys[role] = keygen(dir);
      console.log(`    ${role}: generated`);
    }
  }

  // Step 2: faucet prompt
  console.log("\n[2] Fund these accounts at https://testnet.cspr.live/tools/faucet");
  for (const role of roles) {
    const hexFile = path.join(KEYS_DIR, role, "public_key_hex");
    const hex = fs.existsSync(hexFile)
      ? fs.readFileSync(hexFile, "utf8").trim()
      : keys[role].publicKey.toHex();
    console.log(`    ${role.padEnd(20)}  ${hex}`);
  }
  const readline = require("readline");
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  await new Promise(r => rl.question("\n    Press ENTER when all accounts are funded…", () => { rl.close(); r(); }));

  const deployer = keys["fund_agent"];

  // Step 3: check wasm files
  const wasms = ["OracleRegistry","DataMarket","FundVault","Governance"];
  for (const w of wasms) {
    const p = path.join(WASM_DIR, `${w}.wasm`);
    if (!fs.existsSync(p)) {
      console.error(`\nMissing: ${p}\nRun: bash scripts/build_contracts.sh`);
      process.exit(1);
    }
  }
  console.log("\n[3] WASM files OK");

  // Step 4-7: deploy contracts
  console.log("\n[4] Deploying OracleRegistry…");
  let h = await installWasm(deployer, path.join(WASM_DIR, "OracleRegistry.wasm"), []);
  console.log(`    deploy: ${h}\n    ${EXPLORER}/deploy/${h}`);
  await waitForDeploy(h);
  const registryHash = await extractContractHash(h);
  console.log(`    hash: ${registryHash}`);

  console.log("\n[5] Deploying DataMarket…");
  h = await installWasm(deployer, path.join(WASM_DIR, "DataMarket.wasm"), [
    `registry_hash:string=${registryHash}`,
    "fee_bps:u32=250",
  ]);
  console.log(`    deploy: ${h}\n    ${EXPLORER}/deploy/${h}`);
  await waitForDeploy(h);
  const marketHash = await extractContractHash(h);
  console.log(`    hash: ${marketHash}`);

  console.log("\n[5b] Wiring OracleRegistry.set_market…");
  h = await callContract(deployer, registryHash, "set_market",
      [`market:string=${marketHash}`]);
  await waitForDeploy(h);
  console.log(`    wired ✓  ${EXPLORER}/deploy/${h}`);

  console.log("\n[6] Deploying FundVault…");
  // Use SDK's accountHashStr which correctly applies Blake2b (not sha256).
  const fundAcct = keys["fund_agent"].publicKey.toAccountHashStr();   // "account-hash-<hex>"
  const riskAcctHash = keys["risk_agent"].publicKey.toAccountHashStr();
  h = await installWasm(deployer, path.join(WASM_DIR, "FundVault.wasm"), [
    `operator:string=${fundAcct}`,
    "governance_hash:string=pending",
  ]);
  console.log(`    deploy: ${h}\n    ${EXPLORER}/deploy/${h}`);
  await waitForDeploy(h);
  const vaultHash = await extractContractHash(h);
  console.log(`    hash: ${vaultHash}`);

  console.log("\n[7] Deploying Governance…");
  const riskAcct  = riskAcctHash;
  h = await installWasm(deployer, path.join(WASM_DIR, "Governance.wasm"), [
    `proposer:string=${fundAcct}`,
    `risk_agent:string=${riskAcct}`,
    "veto_window_ms:u64=90000",
  ]);
  console.log(`    deploy: ${h}\n    ${EXPLORER}/deploy/${h}`);
  await waitForDeploy(h);
  const govHash = await extractContractHash(h);
  console.log(`    hash: ${govHash}`);

  // Step 8: write testnet.env
  const envContent = [
    `ORACLE_TBILL_KEY=${path.join(KEYS_DIR,"oracle_tbill","secret_key.pem")}`,
    `ORACLE_GOLD_KEY=${path.join(KEYS_DIR,"oracle_gold","secret_key.pem")}`,
    `ORACLE_REINDEX_KEY=${path.join(KEYS_DIR,"oracle_reindex","secret_key.pem")}`,
    `FUND_AGENT_KEY=${path.join(KEYS_DIR,"fund_agent","secret_key.pem")}`,
    `RISK_AGENT_KEY=${path.join(KEYS_DIR,"risk_agent","secret_key.pem")}`,
    `REGISTRY_HASH=${registryHash}`,
    `MARKET_HASH=${marketHash}`,
    `VAULT_HASH=${vaultHash}`,
    `GOV_HASH=${govHash}`,
  ].join("\n") + "\n";
  fs.writeFileSync(path.join(AGENTS, "testnet.env"), envContent);
  console.log("\n[8] agents/testnet.env written ✓");

  console.log("\n" + "═".repeat(50));
  console.log("  DEPLOYMENT COMPLETE");
  console.log("═".repeat(50));
  console.log(`  OracleRegistry: ${EXPLORER}/contract/${registryHash}`);
  console.log(`  DataMarket    : ${EXPLORER}/contract/${marketHash}`);
  console.log(`  FundVault     : ${EXPLORER}/contract/${vaultHash}`);
  console.log(`  Governance    : ${EXPLORER}/contract/${govHash}`);
  console.log("\n  Next: export HELIOS_MODE=testnet && python3 scripts/testnet_round.py --rounds 3\n");
}

// ── CLI ────────────────────────────────────────────────────────────────────────

const [,, cmd, ...rest] = process.argv;

const argMap = {};
let   lastFlag = null;
for (const a of rest) {
  if (a.startsWith("--")) { lastFlag = a.slice(2); argMap[lastFlag] = []; }
  else if (lastFlag)       { argMap[lastFlag].push(a); }
}
const flag  = k => argMap[k]?.[0];
const flags = k => argMap[k] || [];

(async () => {
  switch (cmd) {

    case "keygen": {
      const roles = ["oracle_tbill","oracle_gold","oracle_reindex","fund_agent","risk_agent"];
      for (const role of roles) {
        const dir = path.join(KEYS_DIR, role);
        if (fs.existsSync(path.join(dir, "secret_key.pem"))) {
          console.log(`${role}: already exists`); continue;
        }
        keygen(dir);
        console.log(`${role}: generated → ${dir}/`);
      }
      console.log(`\nFund at: https://testnet.cspr.live/tools/faucet`);
      break;
    }

    case "status": {
      const { CasperServiceByJsonRPC } = sdk();
      const client = new CasperServiceByJsonRPC(NODE_URL);
      try {
        const info = await client.getStatus();
        console.log(`✓ ${NODE_URL}`);
        console.log(`  chain: ${info.chainspec_name}`);
        console.log(`  peers: ${info.peers?.length ?? "?"}`);
      } catch (e) {
        console.error(`✗ ${NODE_URL}: ${e.message}`); process.exit(1);
      }
      break;
    }

    case "install": {
      const kp = loadKey(flag("key"));
      console.log(`Deployer: ${kp.publicKey.toHex()}`);
      const wasmPath = flag("wasm");
      console.log(`WASM: ${wasmPath} (${fs.statSync(wasmPath).size.toLocaleString()} bytes)`);
      const dh = await installWasm(kp, wasmPath, flags("args"),
                                   flag("payment") || GAS_INSTALL);
      console.log(`Deploy   : ${dh}`);
      console.log(`Explorer : ${EXPLORER}/deploy/${dh}`);
      if ("wait" in argMap) await waitForDeploy(dh);
      break;
    }

    case "call": {
      const kp = loadKey(flag("key"));
      const dh = await callContract(kp, flag("contract"), flag("entry-point"),
                                    flags("args"), flag("payment") || GAS_CALL);
      console.log(`Deploy   : ${dh}`);
      console.log(`Explorer : ${EXPLORER}/deploy/${dh}`);
      if ("wait" in argMap) await waitForDeploy(dh);
      break;
    }

    case "deploy-all":
      await deployAll();
      break;

    case "wait":
      await waitForDeploy(rest[0]);
      break;

    default:
      console.log(`Commands: keygen | status | install | call | wait | deploy-all`);
      console.log(`Run with --help on any command for details.`);
  }
})().catch(e => { console.error("Error:", e.message); process.exit(1); });
